ca/ca.go

```diff
@@ -112,50 +118,57 @@ func (ca *CAImpl) makeRootCert(
 	return newCert, nil
 }
 
-func (ca *CAImpl) newRootIssuer() error {
+func (chain *chain) newRootIssuer(ca *CAImpl) error {
 	// Make a root private key
 	rk, err := makeKey()
 	if err != nil {
 		return err
 	}
 	// Make a self-signed root certificate
-	rc, err := ca.makeRootCert(rk, rootCAPrefix, nil)
+	rc, err := chain.makeRootCert(ca, rk, rootCAPrefix, nil)
 	if err != nil {
 		return err
 	}
 
-	ca.root = &issuer{
+	chain.root = &issuer{
 		key:  rk,
 		cert: rc,
 	}
 	ca.log.Printf("Generated new root issuer with serial %s\n", rc.ID)
 	return nil
 }
 
-func (ca *CAImpl) newIntermediateIssuer() error {
-	if ca.root == nil {
+func (chain *chain) newIntermediateIssuer(ca *CAImpl, ik crypto.Signer) error {
+	if chain.root == nil {
 		return fmt.Errorf("newIntermediateIssuer() called before newRootIssuer()")
 	}
 
-	// Make an intermediate private key
-	ik, err := makeKey()
-	if err != nil {
-		return err
-	}
-
 	// Make an intermediate certificate with the root issuer
-	ic, err := ca.makeRootCert(ik, intermediateCAPrefix, ca.root)
+	ic, err := chain.makeRootCert(ca, ik, intermediateCAPrefix, chain.root)
 	if err != nil {
 		return err
 	}
-	ca.intermediate = &issuer{
+	chain.intermediate = &issuer{
 		key:  ik,
 		cert: ic,
 	}
 	ca.log.Printf("Generated new intermediate issuer with serial %s\n", ic.ID)
 	return nil
 }
 
+func newChain(ca *CAImpl, ik crypto.Signer) *chain {
+	chain := &chain{}
+	err := chain.newRootIssuer(ca)
+	if err != nil {
```

Comment:
nit: you can collapse this line and the prev into a single assignment & check `if` expression like:
```
if err := chain.newRootIssuer(ca); err != nil { 
...
```

Reply:
Thanks! Fixed.